### ISSUE TITLE
Memory leak in LSTM with dropout when executed in eager mode

### ISSUE THREAD

**Issue by vsimkus** [tensorflow/tensorflow#35858]
**System information**
- Have I written custom code: Simple LSTM layer with dropouts in eager loop
- OS Platform and Distribution: Linux Ubuntu 18.04
- TensorFlow installed from: pip
- TensorFlow version: v2.0.0 and v2.1.0
- Python version: 3.6.8

**Describe the current behavior**
An LSTM layer with dropout enabled leaks memory in each eager execution. This is likely due to the dropout mask in the RNN layer not being properly released (related to #19671). 

If the dropout arguments are set to 0 then the memory remains constant.

**Describe the expected behavior**
The memory usage should remain constant among batches with dropouts enabled.

**Code to reproduce the issue**
```python
import tensorflow as tf

layer = tf.keras.layers.LSTM(64, return_sequences=True, 
                             dropout=0.4, 
                             recurrent_dropout=0.2)

while True:
    X = tf.random.uniform((500, 100, 64))
    pred = layer(X, training=True)
```

**Other info / logs**
Memory usage when dropouts are enabled taken with memory_profiler
![memory_leak](https://user-images.githubusercontent.com/5730052/72333667-70a78600-36bc-11ea-9ac7-7bc71ec2468d.png)

Memory usage when dropouts are disabled
![no_memory_leak](https://user-images.githubusercontent.com/5730052/72333733-95036280-36bc-11ea-982c-4621ec5776da.png)

**Workaround**
A working workaround is to set `tf.random.set_seed(seed)` in each iteration, which seems to clear the cached masks, but it can seriously affect the experiments.

**Comment by amahendrakar** [comment]
@vsimkus,
In order to expedite the trouble-shooting process, could you please provide the complete code to reproduce the issue and to visualize the memory usage. Thanks!

**Comment by vsimkus** [comment]
Just run the script above with [`memory_profiler`](https://pypi.org/project/memory-profiler/#time-based-memory-usage). 
```bash
mprof run python memory_leak_example.py # Run the script with memory_profiler
mprof plot # Visualise memory usage
```

**Comment by qlzh727** [comment]
With tf-nightly, the memory footprint becomes stable after 5 epochs and doesn't increase afterwards.

Could you confirm this is still the case for you for nightly version?

**Comment by qlzh727** [comment]
I think the issue is fixed recently the adding the kernel cache size limit. Closing this bug for now.

**Comment by vsimkus** [comment]
Thanks. With `tf-nightly` the memory usage indeed does not blow up, although it is quite irregular and does not seem to stabilise.
![memory_leak2](https://user-images.githubusercontent.com/5730052/72621388-7530a080-3941-11ea-9b75-dceb446d862d.png)
It would be good to understand the behaviour.

**Comment by johnght** [comment]
Again, confirm. This is a tricky bug... whenever **implicit dropout of lstm** is activated. Using a separate tf.keras.layers.Dropout is fine.
Btw, **in eager mode** is tricky too, there are many memory leakages since tf2.0. e.g. tf.keras.layers.TimeDistributed, Model.fit_generator... Probably due to the introduction of autograph and implicit eager execution in tf2.0.
